File: src/day01/main.rs
use itertools::Itertools;
use shared::puzzle_input;

fn elf_sum(elf: &str) -> usize {
    elf.split('\n')
        .map(|s| s.trim().parse::<usize>().unwrap())
        .sum()
}
pub fn main() {
    let sorted_totals = puzzle_input!()
        .trim()
        .split("\n\n")
        .map(elf_sum)
        .sorted()
        .rev();
    let p1 = sorted_totals.clone().max().unwrap();
    let p2: usize = sorted_totals.take(3).sum();
    println!("part 1: {}\npart 2: {}", p1, p2);
}

File: src/day01/main.exs
blocks =
  File.read!("puzzle_input/day01")
  |> String.trim()
  |> String.split(~r/\R\R/, trim: True)

[a, b, c | _] =
  for b <- blocks do
    for line <- String.split(b, ~r/\R/, trim: True) do
      {x, _} = Integer.parse(line)
      x
    end
    |> Enum.sum()
  end
  |> Enum.sort(:desc)

IO.puts("part 1: #{a}")
IO.puts("part 2: #{a + b + c}")
